### ISSUE TITLE
Data Export XML option does not work  (export/userdata.xml file not found)

### ISSUE THREAD

**Issue by Alys** [HabitRPG/habitica#10100]
In the [Data Export](https://habitica.com/user/settings/data-export) screen, the "Export User Data (XML)" option can't be used. You see this message (wording depends on browser):

`can't find the file at https://habitica.com/export/userdata.xml

The JSON export option does work.

_Thanks to Cyrus Krieger (a4626bcc-2bff-4c39-a1ce-cf015ee1ba50) for reporting this in the Report a Bug guild._

**Comment by Hus274** [comment]
I've been researching this one.

**Comment by Hus274** [comment]
So I've been looking into and debugging this issue for a fair bit now and this is what I've learned. 

The problem is there is an exception thrown when converting the user doc to xml. The exception is
```
Error: charData should not contain characters not allowed in XML
```

But it doesn't show the section of the document that it's errored on or what caused it to error. It looks like @paglias has looked into this or a similar issue and [commented on an issue for this library](https://github.com/michaelkourlas/node-js2xmlparser/issues/61#issuecomment-363520355) to ask that the library provide this additional information.

As the library doesn't provide additional information, I was able to debug the issue by narrowing down the keyspace for the user and trying to convert each individual subsection of the user to XML and determine what was causing the failures. Consistently, the user.pinnedItems is the section of the document that causes the xml conversion to fail. When I omitted the pinnedItems from the user document before converting to xml, the conversion to xml and corresponding test completed successfully 100% of the time.

The strange part of the failures on the pinnedItems section is they don't fail 100% of the time, it seems like they pass one in every five to ten. Some samples that are passing and failing:
```
-- FAILING --

[ { type: 'marketGear',
  path: 'gear.flat.weapon_warrior_0',
  _id: 5aa05fca62072cc163daa407 },
{ type: 'marketGear',
  path: 'gear.flat.armor_warrior_1',
  _id: 5aa05fca62072cc163daa408 },
{ type: 'marketGear',
  path: 'gear.flat.shield_warrior_1',
  _id: 5aa05fca62072cc163daa409 },
{ type: 'marketGear',
  path: 'gear.flat.head_warrior_1',
  _id: 5aa05fca62072cc163daa40a },
{ type: 'potion', path: 'potion', _id: 5aa05fca62072cc163daa40b },
{ type: 'armoire',
  path: 'armoire',
  _id: 5aa05fca62072cc163daa40c } ]

[ { type: 'marketGear',
  path: 'gear.flat.weapon_warrior_0',
  _id: 5aa062c83d7fa2cbea10794e },
{ type: 'marketGear',
  path: 'gear.flat.armor_warrior_1',
  _id: 5aa062c83d7fa2cbea10794f },
{ type: 'marketGear',
  path: 'gear.flat.shield_warrior_1',
  _id: 5aa062c83d7fa2cbea107950 },
{ type: 'marketGear',
  path: 'gear.flat.head_warrior_1',
  _id: 5aa062c83d7fa2cbea107951 },
{ type: 'potion', path: 'potion', _id: 5aa062c83d7fa2cbea107952 },
{ type: 'armoire',
  path: 'armoire',
  _id: 5aa062c83d7fa2cbea107953 } ]

-- WORKING --

[ { type: 'marketGear',
  path: 'gear.flat.weapon_warrior_0',
  _id: 5aa05fdf7f6abac184f7fff7 },
{ type: 'marketGear',
  path: 'gear.flat.armor_warrior_1',
  _id: 5aa05fdf7f6abac184f7fff8 },
{ type: 'marketGear',
  path: 'gear.flat.shield_warrior_1',
  _id: 5aa05fdf7f6abac184f7fff9 },
{ type: 'marketGear',
  path: 'gear.flat.head_warrior_1',
  _id: 5aa05fdf7f6abac184f7fffa },
{ type: 'potion', path: 'potion', _id: 5aa05fdf7f6abac184f7fffb },
{ type: 'armoire',
  path: 'armoire',
  _id: 5aa05fdf7f6abac184f7fffc } ]


[ { type: 'marketGear',
  path: 'gear.flat.weapon_warrior_0',
  _id: 5aa0625256785dcb385c3c28 },
{ type: 'marketGear',
  path: 'gear.flat.armor_warrior_1',
  _id: 5aa0625256785dcb385c3c29 },
{ type: 'marketGear',
  path: 'gear.flat.shield_warrior_1',
  _id: 5aa0625256785dcb385c3c2a },
{ type: 'marketGear',
  path: 'gear.flat.head_warrior_1',
  _id: 5aa0625256785dcb385c3c2b },
{ type: 'potion', path: 'potion', _id: 5aa0625256785dcb385c3c2c },
{ type: 'armoire',
  path: 'armoire',
  _id: 5aa0625256785dcb385c3c2d } ]
```

The only difference I see here between those failing and those passing are different _ids. I've mapped the pinned items to just their ids and tried converting to XML and that is the section of that item that's failing. 

I believe the problem is that the IDs aren't being interpretted as strings and instead being interpretted as a hexadecimal encoding of characters somehow leading to them falling outside of the [accepted unicode range](https://github.com/michaelkourlas/node-xmlcreate/blob/master/src/validate.ts#L28). This is demonstrated by the test when it does pass showing the xml encoding as

```
<user>
    <user>
        <_bsontype>ObjectID</_bsontype>
        <id>Z��ڇzZ߹�ߍ</id>
    </user>
    <user>
        <_bsontype>ObjectID</_bsontype>
        <id>Z��ڇzZ߹�ߎ</id>
    </user>
    <user>
        <_bsontype>ObjectID</_bsontype>
        <id>Z��ڇzZ߹�ߏ</id>
    </user>
    <user>
        <_bsontype>ObjectID</_bsontype>
        <id>Z��ڇzZ߹�ߐ</id>
    </user>
    <user>
        <_bsontype>ObjectID</_bsontype>
        <id>Z��ڇzZ߹�ߑ</id>
    </user>
    <user>
        <_bsontype>ObjectID</_bsontype>
        <id>Z��ڇzZ߹�ߒ</id>
    </user>
</user>
```

And that information isn't helpful to the user. For these reasons, my suggestions for fixing this problem are one of the following:
1. Strip the id field from the user pinned items before converting to xml as they're only helpful in the context of the database which the user doesn't have access to.
2. Convert the id field to a string so it will be parsed correctly by the interpreter.

Both of the above methods fix the problem and make the test pass consistently, I would just like some feedback on which would be the preferred approach.

**Comment by Hus274** [comment]
### TL;DR
Important section from the post above:
>My suggestions for fixing this problem are one of the following:
>1. Strip the id field from the user pinned items before converting to xml as they're only helpful in the context of the database which the user doesn't have access to.
>2. Convert the id field to a string so it will be parsed correctly by the interpreter.

>Both of the above methods fix the problem and make the test pass consistently, I would just like some feedback on which would be the preferred approach.

**Comment by Alys** [comment]
@Hus274 Awesome research, thanks so much!
I think converting the id field to a string would be best. Even though it's not directly useful to the user, I feel a bit uncomfortable about removing it in case there's occasions where it could be of interest (e.g., maybe it would help a user on their local install with debugging). However if another admin says to remove it, I don't feel strongly enough about it to argue. :)

I've marked this as in progress for you but let us know if you'd rather someone else take on the fix.

**Comment by Hus274** [comment]
@Alys thanks! I definitely agree that I'd error on the side of having the data rather than wanting it and not having it. Also this way it's consistent with the JSON export which is good. 

I can definitely implement this as I have both fixes already coded and mostly just need to uncomment the right one and delete the other. 

If I don't hear any objections in the next couple days, I'll go ahead and submit a pull request for changing the ID field to a string. 

Thanks!

**Comment by paglias** [comment]
That's super useful @Hus274 ! I found the way to strip invalid chars from the output but looks like it doesn't work for ObjectIds. I haven't looked in depth but if the id isn't used anywhere it's probably there because it gets added automatically and we should remove it

**Comment by Hus274** [comment]
@paglias I do believe that the ID field is simply the object id that's automatically assigned by mongodb and it's not used anywhere in the code.

**Comment by paglias** [comment]
I think if you add `_id: false` to the array schema it should disappear

**Comment by Hus274** [comment]
Yep, that takes care of it. The tests are passing consistently as well. I'll go ahead and post a pull-request for that change.

**Comment by citrusella** [comment]
Is this supposed to be fixed? I just had it happen.

**Comment by Alys** [comment]
Could be from a different cause. Reopening.

**Comment by SilkAndSlug** [comment]
Also happened to me

**Comment by Alys** [comment]
Another report of this today from (MEGA)uBeauFort (d802744b-3d4d-4950-9a75-c8d384ec231f)

**Comment by shanaqui** [comment]
Another report today from Steve (5d15b440-f113-41f4-b102-5061d25b9fed), I think, with the following error:

> This site can’t be reached
> The web page at https://habitica.com/export/userdata.xml might be temporarily down or it may have moved permanently to a new web address.

Verified by me.

**Comment by beffymaroo** [comment]
I see that this issue is closed but as of today I am unable to use the XML or JSON data export options - getting the same error reported above.

**Comment by paglias** [comment]
Looking at the logs, the XML errors are still happening so I'm reopening the issue. Regarding the JSON version mentioned by @beffymaroo it looks like sometimes the session isn't passed around correctly: it could be a problem due to using a private browsing window or a plugin. I'd suggest to always click on the JSON link here https://habitica.com/user/settings/data-export instead of accessing it directly
